src/GafferScene/EditScopeAlgo.cpp

```diff
@@ -854,3 +855,210 @@ const Gaffer::GraphComponent *GafferScene::EditScopeAlgo::attributeEditReadOnlyR
+
+	return result;
+}
+
+EditScope::ProcessorRegistration g_setMembershipProcessorRegistration( "SetMembershipEdits", setMembershipProcessor );
+
+}  // namespace
+
+void EditScopeAlgo::setSetMembership( Gaffer::EditScope *scope, const std::string &set, const IECore::PathMatcher &paths, EditScopeAlgo::SetMembership state )
+{
+	Node *processor = scope->acquireProcessor( "SetMembershipEdits" );
+	auto rows = processor->getChild<Spreadsheet::RowsPlug>( "edits" );
+
+	auto row = rows->row( set );
+	if( !row )
+	{
+		row = rows->addRow();
+		row->namePlug()->setValue( set );
+	}
+
+	auto addCell = row->cellsPlug()->getChild<Spreadsheet::CellPlug>( g_addSetColumnIndex );
+	auto removeCell = row->cellsPlug()->getChild<Spreadsheet::CellPlug>( g_removeSetColumnIndex );
+
+	auto addStringPlug = addCell->valuePlug<StringVectorDataPlug>();
+	auto removeStringPlug = removeCell->valuePlug<StringVectorDataPlug>();
+
+	ConstStringVectorDataPtr addData = addStringPlug->getValue();
+	ConstStringVectorDataPtr removeData = removeStringPlug->getValue();
+
+	PathMatcher addPaths( addData->readable().begin(), addData->readable().end() );
+	PathMatcher removePaths( removeData->readable().begin(), removeData->readable().end() );
+
+	if( state == EditScopeAlgo::SetMembership::INHERIT )
+	{
+		if( !addPaths.removePaths( paths ) )
+		{
+			removePaths.removePaths( paths );
+		}
+	}
+	else if( state == EditScopeAlgo::SetMembership::IN )
+	{
+		addPaths.addPaths( paths );
+		removePaths.removePaths( paths );
+	}
+	else if( state == EditScopeAlgo::SetMembership::OUT )
+	{
+		addPaths.removePaths( paths );
+		removePaths.addPaths( paths );
+	}
+
+	StringVectorDataPtr addResult = new StringVectorData;
+	StringVectorDataPtr removeResult = new StringVectorData;
+
+	addPaths.paths( addResult->writable() );
+	removePaths.paths( removeResult->writable() );
+
+	sort( addResult->writable().begin(), addResult->writable().end() );
+	sort( removeResult->writable().begin(), removeResult->writable().end() );
+
+	addStringPlug->setValue( addResult );
+	removeStringPlug->setValue( removeResult );
+}
+
+EditScopeAlgo::SetMembership EditScopeAlgo::getSetMembership( Gaffer::EditScope *scope, const std::string &set, const std::string &path )
+{
+	Node *processor = scope->acquireProcessor( "SetMembershipEdits", /* createIfNecessary */ false );
+	if( !processor )
+	{
+		return EditScopeAlgo::SetMembership::INHERIT;
+	}
+
+	auto rows = processor->getChild<Spreadsheet::RowsPlug>( "edits" );
+
+	auto row = rows->row( set );
+	if( !row )
+	{
+		return EditScopeAlgo::SetMembership::INHERIT;
+	}
+
+	auto addCell = row->cellsPlug()->getChild<Spreadsheet::CellPlug>( g_addSetColumnIndex );
+	auto addStringPlug = addCell->valuePlug<StringVectorDataPlug>();
+	ConstStringVectorDataPtr addData = addStringPlug->getValue();
+	PathMatcher addPaths( addData->readable().begin(), addData->readable().end() );
+
+	if( addPaths.match( path ) == PathMatcher::ExactMatch )
+	{
+		return EditScopeAlgo::SetMembership::IN;
```

Comment:
Since the user can directly edit `add` and `remove` paths via the NodeEditor, it's possible for the same path to be in both. And since the removing node is below the adding node, here we're returning the wrong result for that case. We need to check for removals first, then additions. Or reorder the nodes (I can't think of a good rationale for one order or the other though).

Reply:
Fixed in 9b7dc9b158938d80412f8f0b046ea7751ee97d17